New warning -Wimplicit-fallthrough of GCC 7

There are new warnings in the testsuite:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-16.shtml, the two platforms named `Ubuntu-latest-GCC6`:
```
../test/CGAL_ipelets_Demo/multi_delaunay.cpp:141:7: warning: this statement may fall through [-Wimplicit-fallthrough]
../test/CGAL_ipelets_Demo/multi_delaunay.cpp:174:7: warning: this statement may fall through [-Wimplicit-fallthrough]
../test/CGAL_ipelets_Demo/multi_delaunay.cpp:196:9: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/Algebraic_kernel_d/Curve_pair_analysis_2.h:774:17: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/Delaunay_triangulation_2.h:698:5: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/ImageIO/gis_impl.h:271:24: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/ImageIO/recline_impl.h:255:18: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/ImageIO/recline_impl.h:430:17: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/ImageIO/recline_impl.h:447:18: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/Intersections_3/intersection_3_1_impl.h:1071:9: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/Intersections_3/intersection_3_1_impl.h:1105:13: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/QP_models.h:1450:2: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/Regular_triangulation_2.h:1250:6: warning: this statement may fall through [-Wimplicit-fallthrough]
../include/CGAL/Triangulation_data_structure_3.h:3390:7: warning: this statement may fall through [-Wimplicit-fallthrough]
../src/CGAL_Qt5/../../include/CGAL/Qt/GraphicsViewNavigation_impl.h:117:21: warning: this statement may fall through [-Wimplicit-fallthrough]
```

--- a/Algebraic_kernel_d/include/CGAL/Algebraic_kernel_d/Curve_pair_analysis_2.h
+++ b/Algebraic_kernel_d/include/CGAL/Algebraic_kernel_d/Curve_pair_analysis_2.h
@@ -774,7 +774,8 @@ private:
                 if(event_x(i).is_rational()) {
                     return create_event_slice_at_rational(i);
                 }
-                // FALL INTO NEXT CASE
+
+                CGAL_FALLTHROUGH;
             }
             case(CGAL::SHEAR_STRATEGY): {
                 return create_event_slice_with_shear(i);

--- a/CGAL_ImageIO/include/CGAL/ImageIO/gis_impl.h
+++ b/CGAL_ImageIO/include/CGAL/ImageIO/gis_impl.h
@@ -268,8 +268,8 @@ int readGisHeader( const char* name,_image* im)
 
   status = (int)iss.str().length();
   switch ( status ) {
-  case 2 :    im->zdim = 1;
-  case 3 :    im->vdim = 1;
+  case 2 :    im->zdim = 1; CGAL_FALLTHROUGH;
+  case 3 :    im->vdim = 1; CGAL_FALLTHROUGH;
   case 4 :    break;
   default :
     fprintf( stderr, "readGisHeader: unable to read dimensions in '%s'\n", name );

--- a/CGAL_ImageIO/include/CGAL/ImageIO/recline_impl.h
+++ b/CGAL_ImageIO/include/CGAL/ImageIO/recline_impl.h
@@ -253,6 +253,7 @@ RFcoefficientType * InitRecursiveCoefficients( double x,
 	fprintf( stderr, "%s: switch to default coefficients (smoothing).\n", proc );
       }
       derivative = DERIVATIVE_0;
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_0 :
       a0     =  1.68;
       omega0 =  0.6318;
@@ -264,6 +265,7 @@ RFcoefficientType * InitRecursiveCoefficients( double x,
       b1     =  1.723;
       break;
     case DERIVATIVE_1 :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_1_CONTOURS :
       a0     =  -0.6472;
       omega0 =  0.6719;
@@ -293,6 +295,7 @@ RFcoefficientType * InitRecursiveCoefficients( double x,
     /*--- normalisation ---*/
     switch ( derivative ) {
     default :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_0 :
       sumA  = 2.0 * a1 * exp( b0 ) * cos0 * cos0 - a0 * sin0 * exp( 2.0 * b0 );
       sumA += a0 * sin0 - 2.0 * a1 * exp( b0 );
@@ -401,7 +404,9 @@ RFcoefficientType * InitRecursiveCoefficients( double x,
     
     switch ( derivative ) {
     default :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_0 :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_2 :
       RFC->sn1 =   RFC->sp1 - RFC->sd1 * RFC->sp0;
       RFC->sn2 =   RFC->sp2 - RFC->sd2 * RFC->sp0;
@@ -409,7 +414,9 @@ RFcoefficientType * InitRecursiveCoefficients( double x,
       RFC->sn4 = - RFC->sd4 * RFC->sp0;
       break;
     case DERIVATIVE_1 :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_1_CONTOURS :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_3 :
       RFC->sn1 = - RFC->sp1 + RFC->sd1 * RFC->sp0;
       RFC->sn2 = - RFC->sp2 + RFC->sd2 * RFC->sp0;
@@ -445,6 +452,7 @@ RFcoefficientType * InitRecursiveCoefficients( double x,
 	fprintf( stderr, "%s: switch to default coefficients (smoothing).\n", proc );
       }
       derivative = DERIVATIVE_0;
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_0 :
       RFC->sp0 = (1.0 - ex) * (1.0 - ex) / (1.0 + 2.0 * x * ex - ex * ex);
       RFC->sp1 = RFC->sp0 * (x - 1.0) * ex;
@@ -535,6 +543,7 @@ int RecursiveFilter1D( RFcoefficientType *RFC,
       fprintf( stderr, "%s: unknown type of recursive filter.\n", proc );
     return( EXIT_ON_FAILURE );
   case GAUSSIAN_FIDRICH :
+    CGAL_FALLTHROUGH;
   case GAUSSIAN_DERICHE :
     /*--- filtrage generique d'ordre 4 ---*/
     rp0 = RFC->sp0;   rp1 = RFC->sp1;   rp2 = RFC->sp2;   rp3 = RFC->sp3;
@@ -583,7 +592,9 @@ int RecursiveFilter1D( RFcoefficientType *RFC,
     
     switch( RFC->derivative ) {
     default :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_0 :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_2 :
 
       rp0 = RFC->sp0;   rp1 = RFC->sp1;
@@ -619,6 +630,7 @@ int RecursiveFilter1D( RFcoefficientType *RFC,
       break;
       
     case DERIVATIVE_1 :
+      CGAL_FALLTHROUGH;
     case DERIVATIVE_1_CONTOURS :
       rp1 = RFC->sp1;
       rn1 = RFC->sn1;

--- a/CGAL_ImageIO/include/CGAL/ImageIO/recline_impl.h
+++ b/CGAL_ImageIO/include/CGAL/ImageIO/recline_impl.h
@@ -435,6 +435,7 @@ RFcoefficientType * InitRecursiveCoefficients( double x,
       fprintf( stderr, "%s: switch to default recursive filter (Deriche's filters).\n", proc );
     }
     type_filter = ALPHA_DERICHE;
+    CGAL_FALLTHROUGH;
   case ALPHA_DERICHE :
 
     if ( (x < 0.1) || (x > 1.9) ) {

--- a/CGAL_ipelets/demo/CGAL_ipelets/multi_delaunay.cpp
+++ b/CGAL_ipelets/demo/CGAL_ipelets/multi_delaunay.cpp
@@ -141,8 +141,8 @@ void MdelaunayIpelet::protected_run(int fn)
           draw_in_ipe(rt);
           break;
         }
-        CGAL_FALLTHROUGH;
       }
+      CGAL_FALLTHROUGH;
     case 3://Delaunay and Voronoi of order n-1
       CGAL_FALLTHROUGH;
     case 8:
@@ -176,8 +176,8 @@ void MdelaunayIpelet::protected_run(int fn)
           draw_in_ipe(rt);
           break;
         }
-        CGAL_FALLTHROUGH;
       }
+      CGAL_FALLTHROUGH;
     case 4:
       CGAL_FALLTHROUGH;
     case 9://k-th Delauney and Voronoi
@@ -200,8 +200,8 @@ void MdelaunayIpelet::protected_run(int fn)
             draw_in_ipe(rt);
             break;
           }
-          CGAL_FALLTHROUGH;
         }
+        CGAL_FALLTHROUGH;
     case 5://Draw Voronoi diagrams
       if(fn==5) draw_dual_in_ipe(dt,bbox);
       if(fn==6) draw_dual_in_ipe(rti,bbox);

--- a/CGAL_ipelets/demo/CGAL_ipelets/multi_delaunay.cpp
+++ b/CGAL_ipelets/demo/CGAL_ipelets/multi_delaunay.cpp
@@ -98,8 +98,11 @@ void MdelaunayIpelet::protected_run(int fn)
       draw_in_ipe(dt);
       break;
     case 1:
+      CGAL_FALLTHROUGH;
     case 6:
+      CGAL_FALLTHROUGH;
     case 2:
+      CGAL_FALLTHROUGH;
     case 7:        //Delaunay and Voronoi for 2nd-3rd order
       for (Delaunay::Finite_edges_iterator it=dt.finite_edges_begin();it!=dt.finite_edges_end();++it){
           Point_2 pt0=it->first->vertex(Delaunay::cw(it->second))->point();
@@ -138,8 +141,10 @@ void MdelaunayIpelet::protected_run(int fn)
           draw_in_ipe(rt);
           break;
         }
+        CGAL_FALLTHROUGH;
       }
     case 3://Delaunay and Voronoi of order n-1
+      CGAL_FALLTHROUGH;
     case 8:
       if(fn==3 ||fn==8){
         int order = pt_list.size()-1;
@@ -171,8 +176,10 @@ void MdelaunayIpelet::protected_run(int fn)
           draw_in_ipe(rt);
           break;
         }
+        CGAL_FALLTHROUGH;
       }
     case 4:
+      CGAL_FALLTHROUGH;
     case 9://k-th Delauney and Voronoi
         if(fn==4 ||fn==9){
           int order;
@@ -193,6 +200,7 @@ void MdelaunayIpelet::protected_run(int fn)
             draw_in_ipe(rt);
             break;
           }
+          CGAL_FALLTHROUGH;
         }
     case 5://Draw Voronoi diagrams
       if(fn==5) draw_dual_in_ipe(dt,bbox);

--- a/GraphicsView/include/CGAL/Qt/GraphicsViewNavigation_impl.h
+++ b/GraphicsView/include/CGAL/Qt/GraphicsViewNavigation_impl.h
@@ -115,6 +115,7 @@ namespace Qt {
       case ::Qt::Key_Control:
         cursor_backup = v->cursor();
         v->setCursor(::Qt::CrossCursor);
+        CGAL_FALLTHROUGH;
       default:
         return false;
       }

--- a/Installation/include/CGAL/config.h
+++ b/Installation/include/CGAL/config.h
@@ -563,11 +563,11 @@ typedef const void * Nullptr_t;   // Anticipate C++0x's std::nullptr_t
 //   https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
 #if __has_cpp_attribute(fallthrough)
 #  define CGAL_FALLTHROUGH [[fallthrough]]
-#elseif __has_cpp_attribute(gnu::fallthrough)
+#elif __has_cpp_attribute(gnu::fallthrough)
 #  define CGAL_FALLTHROUGH [[gnu::fallthrough]]
-#elseif __has_cpp_attribute(clang::fallthrough)
+#elif __has_cpp_attribute(clang::fallthrough)
 #  define CGAL_FALLTHROUGH [[clang::fallthrough]]
-#elseif __has_attribute(fallthrough)
+#elif __has_attribute(fallthrough)
 #  define CGAL_FALLTHROUGH __attribute__ ((fallthrough))
 #else
 #  define CGAL_FALLTHROUGH while(false){}

--- a/Installation/include/CGAL/config.h
+++ b/Installation/include/CGAL/config.h
@@ -567,7 +567,7 @@ typedef const void * Nullptr_t;   // Anticipate C++0x's std::nullptr_t
 #  define CGAL_FALLTHROUGH [[gnu::fallthrough]]
 #elif __has_cpp_attribute(clang::fallthrough)
 #  define CGAL_FALLTHROUGH [[clang::fallthrough]]
-#elif __has_attribute(fallthrough)
+#elif __has_attribute(fallthrough) && ! __clang__
 #  define CGAL_FALLTHROUGH __attribute__ ((fallthrough))
 #else
 #  define CGAL_FALLTHROUGH while(false){}

--- a/Installation/include/CGAL/config.h
+++ b/Installation/include/CGAL/config.h
@@ -409,6 +409,7 @@ using std::max;
 // Macros to detect features of clang. We define them for the other
 // compilers.
 // See http://clang.llvm.org/docs/LanguageExtensions.html
+// See also http://en.cppreference.com/w/cpp/experimental/feature_test
 #ifndef __has_feature
   #define __has_feature(x) 0  // Compatibility with non-clang compilers.
 #endif
@@ -423,6 +424,9 @@ using std::max;
 #endif
 #ifndef __has_attribute
   #define __has_attribute(x) 0  // Compatibility with non-clang compilers.
+#endif
+#ifndef __has_cpp_attribute
+  #define __has_cpp_attribute(x) 0  // Compatibility with non-supporting compilers.
 #endif
 #ifndef __has_warning
   #define __has_warning(x) 0  // Compatibility with non-clang compilers.
@@ -552,27 +556,21 @@ typedef const void * Nullptr_t;   // Anticipate C++0x's std::nullptr_t
 #define CGAL_NOEXCEPT(x)
 #endif
 
-// Very preliminary support for [[fallthrough]]
-
-// For boost >= 1.54 we can use BOOST_FALL_THROUGH
-// http://www.boost.org/doc/libs/1_59_0/libs/config/doc/html/boost_config/boost_macro_reference.html
-// says:
-// When compiled with Clang >3.2 in C++11 mode, the BOOST_FALLTHROUGH macro is 
-// expanded to [[clang::fallthrough]] attribute
-
-
-// For g++:  
-// https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
-// says:
-// GCC provides an attribute, __attribute__ ((fallthrough))
-// C++17 provides a standard way to suppress the -Wimplicit-fallthrough warning using [[fallthrough]]; 
-// instead of the GNU attribute. 
-// In C++11 or C++14 users can use [[gnu::fallthrough]];
-
-#ifdef CGAL_USE_GNU_FALLTHROUGH
-#define CGAL_FALLTHROUGH [[gnu::fallthrough]]
+// The fallthrough attribute
+// See for clang:
+//   http://clang.llvm.org/docs/AttributeReference.html#statement-attributes
+// See for gcc:
+//   https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
+#if __has_cpp_attribute(fallthrough)
+#  define CGAL_FALLTHROUGH [[fallthrough]]
+#elseif __has_cpp_attribute(gnu::fallthrough)
+#  define CGAL_FALLTHROUGH [[gnu::fallthrough]]
+#elseif __has_cpp_attribute(clang::fallthrough)
+#  define CGAL_FALLTHROUGH [[clang::fallthrough]]
+#elseif __has_attribute(fallthrough)
+#  define CGAL_FALLTHROUGH __attribute__ ((fallthrough))
 #else
-#define CGAL_FALLTHROUGH while(false){}
+#  define CGAL_FALLTHROUGH while(false){}
 #endif
 
 // https://svn.boost.org/trac/boost/ticket/2839

--- a/Installation/include/CGAL/config.h
+++ b/Installation/include/CGAL/config.h
@@ -552,6 +552,29 @@ typedef const void * Nullptr_t;   // Anticipate C++0x's std::nullptr_t
 #define CGAL_NOEXCEPT(x)
 #endif
 
+// Very preliminary support for [[fallthrough]]
+
+// For boost >= 1.54 we can use BOOST_FALL_THROUGH
+// http://www.boost.org/doc/libs/1_59_0/libs/config/doc/html/boost_config/boost_macro_reference.html
+// says:
+// When compiled with Clang >3.2 in C++11 mode, the BOOST_FALLTHROUGH macro is 
+// expanded to [[clang::fallthrough]] attribute
+
+
+// For g++:  
+// https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html
+// says:
+// GCC provides an attribute, __attribute__ ((fallthrough))
+// C++17 provides a standard way to suppress the -Wimplicit-fallthrough warning using [[fallthrough]]; 
+// instead of the GNU attribute. 
+// In C++11 or C++14 users can use [[gnu::fallthrough]];
+
+#ifdef CGAL_USE_GNU_FALLTHROUGH
+#define CGAL_FALLTHROUGH [[gnu::fallthrough]]
+#else
+#define CGAL_FALLTHROUGH while(false){}
+#endif
+
 // https://svn.boost.org/trac/boost/ticket/2839
 #if defined(BOOST_MSVC) && BOOST_VERSION < 105600
 #define CGAL_CFG_BOOST_VARIANT_SWAP_BUG 1

--- a/Intersections_3/include/CGAL/Intersections_3/intersection_3_1_impl.h
+++ b/Intersections_3/include/CGAL/Intersections_3/intersection_3_1_impl.h
@@ -1073,7 +1073,7 @@ intersection(const typename K::Plane_3 &plane,
             return intersection_return<typename K::Intersect_3, typename K::Plane_3, typename K::Segment_3>(target);
         case ON_POSITIVE_SIDE:
             return intersection_return<typename K::Intersect_3, typename K::Plane_3, typename K::Segment_3>();
-        case ON_NEGATIVE_SIDE:
+        default: // ON_NEGATIVE_SIDE:
           { 
             // intersection object should be a point, but rounding errors 
             // could lead to:

--- a/Mesh_3/include/CGAL/Mesh_3/Robust_intersection_traits_3.h
+++ b/Mesh_3/include/CGAL/Mesh_3/Robust_intersection_traits_3.h
@@ -391,7 +391,8 @@ ts_intersection(const typename K::Triangle_3 &t,
 #else
           return make_object(p);
 #endif
-        }
+        } else
+          return result_type();
       case NEGATIVE:
         // q sees the triangle in clockwise order
         if(orientation(p,q,a,b) != POSITIVE
@@ -403,7 +404,8 @@ ts_intersection(const typename K::Triangle_3 &t,
 #else
           return make_object(p);
 #endif
-        }
+        } else
+          return result_type();
       case COPLANAR:
         // the segment is coplanar with the triangle's supporting plane
         // we test whether the segment intersects the triangle in the common

--- a/QP_solver/include/CGAL/QP_models.h
+++ b/QP_solver/include/CGAL/QP_models.h
@@ -1448,7 +1448,7 @@ private:
       switch (type) {
       case FX:
 	this->set_u (var_index, true, val);
-	// no break here!!
+	CGAL_FALLTHROUGH;
       case LO:
 	this->set_l (var_index, true, val);
 	break;

--- a/TDS_3/include/CGAL/Triangulation_data_structure_3.h
+++ b/TDS_3/include/CGAL/Triangulation_data_structure_3.h
@@ -3393,7 +3393,7 @@ is_valid(bool verbose, int level ) const
         CGAL_triangulation_assertion(false);
         return false;
       }
-      // no break; continue
+      CGAL_FALLTHROUGH;
     }
   case -1:
     {

--- a/Triangulation_2/include/CGAL/Delaunay_triangulation_2.h
+++ b/Triangulation_2/include/CGAL/Delaunay_triangulation_2.h
@@ -695,7 +695,7 @@ nearest_vertex(const Point  &p, Face_handle f) const
   switch (this->dimension()) {
   case 0:
     if (this->number_of_vertices() == 0) return Vertex_handle();
-    if (this->number_of_vertices() == 1) return this->finite_vertex();
+    return this->finite_vertex();
     //break;
   case 1:
     return nearest_vertex_1D(p);

--- a/Triangulation_2/include/CGAL/Regular_triangulation_2.h
+++ b/Triangulation_2/include/CGAL/Regular_triangulation_2.h
@@ -1245,7 +1245,7 @@ insert(const Weighted_point &p, Locate_type lt, Face_handle loc, int li)
 	      regularize (v);
 	      return v;
 	      
-	    case ON_ORIENTED_BOUNDARY:
+	    default: // that is ON_ORIENTED_BOUNDARY:
 	      return vv;
 	    }
         }

